### DISCUSSION TITLE
Is there a way to use Pydantic with pandas DataFrames?

### DISCUSSION THREAD

**Discussion by MihanixA** [pydantic/pydantic#2690] [Question]
A have a function returning a DataFrame, and I wonder if I could use type hints to specify columns and their structure
Example 
```python
def f() -> DataFrame:
    return DataFrame([['a', 1], ['b', None]], columns=['label', 'number'])
```
Desired type hints, something like this:
```python
FType = DataFrameType[DataFrameColumnType['label', str], DataFrameColumnType['number', Optional[int]]]

def f() -> FType:
    return DataFrame([['a', 1], ['b', None]], columns=['label', 'number'])
```

**Comment by daviskirk** [comment]
Unfortunately, I don't think there's a way for static type checkers like mypy to understand something like this without a plugin or pandas itself adding typing support for this.

If you're just looking to validate input data directly at runtime, there's probably a way to hack this with pydantic, but there are also a few pandas specific libraries that were inspired by pydantic like [pandera](https://github.com/pandera-dev/pandera) for example.  Does that help?
